diff --git a/DataSourceCloudStack.py b/DataSourceCloudStack_new.py
index 98189c2..d55bafc 100644
--- a/DataSourceCloudStack.py
+++ b/DataSourceCloudStack_new.py
@@ -22,7 +22,7 @@ from struct import pack
 from cloudinit import sources, subp
 from cloudinit import url_helper as uhelp
 from cloudinit import util
-from cloudinit.net import dhcp
+from cloudinit.net import dhcp, find_candidate_nics
 from cloudinit.net.dhcp import NoDHCPLeaseError
 from cloudinit.sources.helpers import ec2
 
@@ -127,6 +127,10 @@ class DataSourceCloudStack(sources.DataSource):
             "%s",
             self.distro.dhcp_client.client_name,
         )
+        for candidate_nic in find_candidate_nics():
+            domain_name = dhcp.networkmanager_get_option_from_leases(candidate_nic, "domain_name")
+            if domain_name:
+                return domain_name
         try:
             latest_lease = self.distro.dhcp_client.get_newest_lease(
                 self.distro.fallback_interface
@@ -315,17 +319,23 @@ def get_vr_address(distro):
             LOG.debug("Found SERVER_ADDRESS '%s' via dhclient", latest_address)
             return latest_address
 
-    with suppress(FileNotFoundError):
-        latest_lease = distro.dhcp_client.get_newest_lease(
-            distro.fallback_interface
-        )
-        if latest_lease:
-            LOG.debug(
-                "Found SERVER_ADDRESS '%s' via ephemeral %s lease ",
-                latest_lease,
-                distro.dhcp_client.client_name,
-            )
-            return latest_lease
+    for candidate_nic in find_candidate_nics():
+        current_lease = dhcp.networkmanager_get_option_from_leases(candidate_nic, "dhcp_server_identifier")
+        if current_lease:
+            LOG.debug("Fount SERVER_ADDRESS '%s' via NetworkManager lease", current_lease)
+            return current_lease
+
+    with suppress(FileNotFoundError, dhcp.NoDHCPLeaseError):
+        for candidate_nic in find_candidate_nics():
+            LOG.debug(f"Checking dhcp server on interface {candidate_nic}")
+            latest_lease = distro.dhcp_client.get_newest_lease(candidate_nic).get("dhcp-server-identifier")
+            if latest_lease:
+                LOG.debug(
+                    "Found SERVER_ADDRESS '%s' via ephemeral %s lease ",
+                    latest_lease,
+                    distro.dhcp_client.client_name,
+                )
+                return latest_lease
 
     # No virtual router found, fallback to default gateway
     LOG.debug("No DHCP found, using default gateway")
